### PR TITLE
BREAKING CHANGE: Update WebApi routes for security and RESTful best practices

### PR DESCRIPTION
**Pull request by glahaye**
### Motivation and Context
The REST routes have grown organically and don't represent the best REST practices.
Also, user ID is sometimes taken from routes instead of verified tokens.
This PR is to address these two points.

### Description
- Changed routes
- Renamed classes accordingly
- Hid chat archive (formally "bot") schema settings from user
- Simplified IContentSafetyService interface and related dependency injection
- Removed ConfigureAwait(false) instances (they have no effect in ASP.Net Core apps)
- Moved isContentSafetyEnabled to ServiceOptions

### Contribution Checklist
- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [Contribution Guidelines](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Review by dehoward — CHANGES_REQUESTED** [review]

File: webapp/src/libs/services/BotService.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ export class BotService extends BaseService {
     public uploadAsync = async (bot: Bot, accessToken: string) => {
         const result = await this.getResponseAsync<IChatSession>(
             {
-                commandPath: 'bot/upload',
+                commandPath: 'chats/archives',
```
[the remark this answers is not in the document]
Comment (crickman) [reply webapp/src/libs/services/BotService.ts:24]: Yes, import/upload isn't currently supported.

https://github.com/microsoft/chat-copilot/issues/342

**Review by crickman — APPROVED** [review]
Checked areas for which I am familiar (memory stuff)...looks good.,

File: webapi/Skills/ChatSkills/SemanticMemoryRetriever.cs [hunk 2]
```diff
@@ -77,7 +77,7 @@ public class SemanticMemoryRetriever
         List<(Citation Citation, Citation.Partition Memory)> relevantMemories = new();
         foreach (var memoryName in this._memoryNames)
         {
-            await SearchMemoryAsync(memoryName).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (alliscode) [reply webapi/Skills/ChatSkills/SemanticMemoryRetriever.cs:80]: ASP.NET Core does not have a SynchronizationContext so ConfigureAwait doesn't do anything, it's basically a no-op.

File: webapi/Controllers/ChatController.cs [hunk 3]
```diff
@@ -90,10 +91,12 @@ public ChatController(ILogger<ChatController> logger, ITelemetryService telemetr
         [FromServices] ChatSessionRepository chatSessionRepository,
         [FromServices] ChatParticipantRepository chatParticipantRepository,
         [FromServices] IAuthInfo authInfo,
-        [FromBody] Ask ask)
+        [FromBody] Ask ask,
+        [FromRoute] Guid chatId)
     {
         this._logger.LogDebug("/chat request received.");
```
[the remark this answers is not in the document]
Comment (TaoChenOSU) [reply webapi/Controllers/ChatController.cs:97]: And should we downgrade this to a trace message?

File: webapi/Controllers/ChatController.cs [hunk 4]
```diff
@@ -270,11 +263,12 @@ public ChatController(ILogger<ChatController> logger, ITelemetryService telemetr
     /// <param name="manifestDomain">The domain of the manifest.</param>
     /// <returns>The plugin's manifest JSON.</returns>
     [HttpGet]
-    [Route("getPluginManifest")]
+    [Route("pluginManifest")]
```
[the remark this answers is not in the document]
Comment (TaoChenOSU) [reply webapi/Controllers/ChatController.cs:266]: This has been moved to a new plugin controller.

File: webapi/Controllers/ChatController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (glahaye) [reply webapi/Controllers/ChatController.cs:97]: Done

File: webapi/Controllers/ChatArchiveController.cs [hunk 5]
```diff
@@ -21,121 +21,113 @@
...
     [ProducesResponseType(StatusCodes.Status404NotFound)]
     [Authorize(Policy = AuthPolicyName.RequireChatParticipant)]
-    public async Task<ActionResult<Bot?>> DownloadAsync(Guid chatId, CancellationToken cancellationToken = default)
+    public async Task<ActionResult<ChatArchive?>> DownloadAsync(Guid chatId, CancellationToken cancellationToken = default)
     {
-        this._logger.LogDebug("Received call to download a bot");
+        this._logger.LogDebug("Received call to download a chat archive");
 
-        var memory = await this.CreateBotAsync(chatId, cancellationToken);
+        var memory = await this.CreateChatArchiveAsync(chatId, cancellationToken);
```
[the remark this answers is not in the document]
Comment (glahaye) [reply webapi/Controllers/ChatArchiveController.cs:77]: Taken in

File: webapi/Controllers/ChatController.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (glahaye) [reply webapi/Controllers/ChatController.cs:266]: Done

File: webapi/Controllers/ServiceInfoController.cs [hunk 6]
```diff
@@ -28,37 +28,41 @@ public class ServiceInfoController : ControllerBase
...
 
     /// <summary>
-    /// Return the memory store type that is configured.
+    /// Return information on running service.
     /// </summary>
-    [Route("serviceOptions")]
+    [Route("info")]
     [HttpGet]
     [ProducesResponseType(StatusCodes.Status200OK)]
     public IActionResult GetServiceOptions()
```
[the remark this answers is not in the document]
Comment (glahaye) [reply webapi/Controllers/ServiceInfoController.cs:55]: Done

File: webapi/Options/ChatArchiveSchemaInfo.cs [hunk 7]
```diff
@@ -5,21 +5,19 @@
...
+public record ChatArchiveSchemaInfo
 {
-    public const string PropertyName = "BotSchema";
-
     /// <summary>
     /// The name of the schema.
     /// </summary>
     [Required, NotEmptyOrWhitespace]
-    public string Name { get; set; } = string.Empty;
+    public string Name { get; init; } = "CopilotChat";
```
[the remark this answers is not in the document]
Comment (glahaye) [reply webapi/Options/ChatArchiveSchemaInfo.cs:16]: This is for serialization of chat archives.  I'm just making the default what we currently use. I don't anticipate this changing soon (or ever, for that matter)

File: webapp/src/libs/services/BotService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (glahaye) [reply webapp/src/libs/services/BotService.ts:24]: The client will be ready when the route does materialize!

File: webapi/Controllers/ChatController.cs [hunk 8]
```diff
@@ -123,10 +127,12 @@ public ChatController(ILogger<ChatController> logger, ITelemetryService telemetr
         [FromServices] ChatSessionRepository chatSessionRepository,
         [FromServices] ChatParticipantRepository chatParticipantRepository,
         [FromServices] IAuthInfo authInfo,
-        [FromBody] ExecutePlanParameters ask)
+        [FromBody] ExecutePlanParameters ask,
```
[the remark this answers is not in the document]
Comment (glahaye) [reply webapi/Controllers/ChatController.cs:130]: Done

File: webapi/Controllers/ChatArchiveController.cs [hunk 9]
```diff
@@ -21,121 +21,113 @@
...
-    /// <param name="kernel">The Semantic Kernel instance.</param>
-    /// <param name="chatId">The chat id to be downloaded.</param>
-    /// <returns>The serialized Bot object of the chat id.</returns>
+    /// <param name="chatId">The ID of chat to be downloaded.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>The serialized chat archive object of the chat id.</returns>
     [HttpGet]
-    [ActionName("DownloadAsync")]
-    [Route("bot/download/{chatId:guid}")]
+    [Route("chats/archives/{chatId:guid}")]
```
[the remark this answers is not in the document]
Comment (glahaye) [reply webapi/Controllers/ChatArchiveController.cs:68]: I agree with @dehoward 

@alliscode although your proposal is interesting, going with what Desmond suggests keeps everything uniform and introduces fewer ways of thinking about our collections.

If everybody is cool with "chats/{chatId:guid}/archive", I'll go with that.

**Review by alliscode — APPROVED** [review]

**Review by dehoward — APPROVED** [review]

**Pull request merged**